adal/src/androidTest/java/com/microsoft/aad/adal/AuthenticationContextTest.java

```diff
@@ -583,6 +590,50 @@ public void testPrompt() throws InterruptedException {
         assertEquals("Prompt param is same", PromptBehavior.REFRESH_SESSION, authenticationRequest2.getPrompt());
     }
 
+    @Test
+    public void testTokenContainsDeviceId() throws InterruptedException, IOException, ParseException, PackageManager.NameNotFoundException {
+        FileMockContext mockContext = new FileMockContext(getInstrumentation().getContext());
+        ITokenCacheStore mockCache = getCacheForRefreshToken(TEST_IDTOKEN_USERID, TEST_IDTOKEN_UPN);
+        final AuthenticationContext context = getAuthenticationContext(mockContext,
+                VALID_AUTHORITY, false, mockCache);
+
+        final MockActivity testActivity = new MockActivity();
+        final CountDownLatch signal = new CountDownLatch(1);
+        testActivity.mSignal = signal;
+        MockAuthenticationCallback callback = new MockAuthenticationCallback(signal);
+        final String response = "{\"id_token\":\""
+                + TEST_IDTOKEN
+                + "\",\"access_token\":\"" + TEST_ACCESS_TOKEN_WITH_DEVICE_ID + "\",\"token_type\":\"Bearer\",\"expires_in\":\"-10\",\"expires_on\":\"1368768616\",\"refresh_token\":\"refresh112\",\"scope\":\"*\",\"deviceid\":\"1234\", \"client_info\":\"" + Util.TEST_CLIENT_INFO + "\"}";
+
+        final HttpURLConnection mockedConnection = Mockito.mock(HttpURLConnection.class);
+        HttpUrlConnectionFactory.setMockedHttpUrlConnection(mockedConnection);
+        Util.prepareMockedUrlConnection(mockedConnection);
+        Mockito.when(mockedConnection.getOutputStream()).thenReturn(Mockito.mock(OutputStream.class));
+        Mockito.when(mockedConnection.getInputStream()).thenReturn(Util.createInputStream(response));
+        Mockito.when(mockedConnection.getInputStream()).thenReturn(Util.createInputStream(response));
+        Mockito.when(mockedConnection.getInputStream()).thenReturn(Util.createInputStream(response));
+        Mockito.when(mockedConnection.getInputStream()).thenReturn(Util.createInputStream(response));
+        Mockito.when(mockedConnection.getResponseCode()).thenReturn(HttpURLConnection.HTTP_OK);
+
+        final PackageManager mockedPackageManager = Mockito.mock(PackageManager.class);
+        final PackageInfo mockPackageInfo = new PackageInfo();
+        mockPackageInfo.setLongVersionCode(Long.MAX_VALUE);
+        Mockito.when(mockedPackageManager.getPackageInfo(Mockito.anyString(), Mockito.anyInt())).thenReturn(mockPackageInfo);
+        mockContext.setMockedPackageManager(mockedPackageManager);
+
+        // Call acquire token which will try refresh token based on cache
+        context.acquireToken(testActivity.getTestActivity(), "resource", "clientid", "redirectUri",
+                TEST_IDTOKEN_UPN, PromptBehavior.Auto, null, "{\"access_token\":{\"deviceid\":{\"essential\":true}}}", callback);
```

Comment:
does it matter what claim is passed, based on the mocking it is going to return the same token right, irrespective of what claims are passed.

Reply:
Yes, the same token will be returned. This is moreso checking that we can pass in the claim and actually complete an acquiretoken regardless of what the result we actually get is.